### DISCUSSION TITLE
Test Promise without trying expect() inside then() callback.

### DISCUSSION THREAD

**Discussion by AlirezaEthDev** [chaijs/chai#1502] [Support] [answered]
As I read [Chai Http](https://www.chaijs.com/plugins/chai-http/), Whenever we need to test a promise by Chai Http, we should try ```expect()``` in a ```then()``` callback function, like this:

```
chai.request(app)
  .put('/user/me')
  .send({ password: '123', confirmPassword: '123' })
  .then(function (res) {
     expect(res).to.have.status(200);
  })
```

My questiona are:
1) Is it possible to try ```expect()``` for a promise in out of ```then()```, For example in another js file that requires this file?
2) How can I do that?

**Comment by keithamus** [comment, marked as answer]
Hey @AlirezaEthDev thanks for the issue. 

The best way to do this is with async/await:

```js
test('the thing', async () => {
  expect(
    await chai.request(app)
//  ^ note the "await"
      .send({ password: '123', confirmPassword: '123' })
  ).to.have.status(200)
})
```

If you can't use async await you could try chai-as-promised:

```js
chai.use(require('chai-as-promised'))
test('the thing', () => {
  return expect(
    chai.request(app)
    .send({ password: '123', confirmPassword: '123' })
  ).to.eventually.have.status(200)
  //   ^ note the "eventually" from chai-as-promised
})
```

**Reply by AlirezaEthDev** [reply]
@keithamus Thanks for answer. Suppose:
```
chai.use(require('chai-as-promised'))
test('the thing', () => {
  return expect(reqTo.postReq()).to.eventually.have.status(200)
  //   ^ note the "eventually" from chai-as-promised
})
```
So that ```reqTo``` is a module (Is in another file) and ```reqTo.postReq()``` runs a thenable function inside itself. As I tried, the above solution doesn't work for that. How can I test ```reqTo.postReq()``` by ```expect()``` here?

Thanks.

**Reply by keithamus** [reply]
Without knowing what `reqTo.postReq()` does I can't give you an answer I'm afraid. I can provide some basic guidelines:

 - `reqTo.postReq()` should return a Promise of the result.
 - `have.status(200)` will test that the result is a Chai Request Test Result (the thing that `chai.request(app)` creates). If it isn't, the test will fail. So `reqTo.postReq()` will need to call `chai.request(app)` internally.

**Reply by AlirezaEthDev** [reply]
Thanks again.
```reqTo``` is:

```
const axios=require("axios");
const myReq={
    username:<username>,
     password: <password>
}

exports.postReq=function(){
    axios.post("http://127.0.0.1:8000/auth/",myReq).then(function(res){
       return res;
    });
}
```
As you see, I'm using Axios for requesting in my main code. According to your reply, I should send requests by ```chai.request(app)...``` instead ```axios.post(...)...``` in my main code, are I right?

**Reply by keithamus** [reply]
If you want to use `chai-http` you'll need to use `chai.request(app)`, yes. If you want to continue to use axios, you can do so without chai-http, but you'll use regular assertions instead of the chai-http ones. If you do an internet search for "chai axios" you will come across a variety of articles about how to write tests in chai using axios.

**Reply by AlirezaEthDev** [reply]
Ok, I search and feedback to you. Yes, I  want to continue to use axios.

**Reply by AlirezaEthDev** [reply]
@keithamus Hi again!
Finally, After some research I could solve the issue by JavaScript Promise object. I don't claim this is the best solution but it works. 

Solution:
1) I modified ```reqTo.js```:
```
const axios=require("axios");
const myReq={
    username:<username>,
     password: <password>
  }

exports.postReq=async function(){
    return await axios.post("http://127.0.0.1:8000/auth/",myReq);
}
```
As you can see, I just changed ```postReq()``` to an async function and removed ```.then()``` after ```axios.post()```.

2) I used this code in the test file that applies JavaScript Promise object:
```
const reqTo=require(".../app/reqTo.js");
const chai=require("chai");
const { describe } = require("mocha");
const expect=chai.expect;

    describe("Test Promise",function(){
        it("Test by new Promise",function(){
                    const promise=new Promise(async function(myResolve,myReject){
                        const result=await reqTo.postReq();
                              if(result.status==200){
                                myResolve(200);
                              }else{
                                myReject("Error: Expected "+result.status+" to be equal to 200");
                              }
                });
               promise.then(
                   function(value){expect(value).to.equal(200);
                       console.log("Resolved: "+value);
                }
               ).catch(function(err){
                   console.log(err);
               });
        })
    });
```

And I it works.

Thanks for your helps and guides!
